### PR TITLE
MAINT: silence warning generated by `spatial.directed_hausdorff`

### LINKED ISSUE #13468: directed_hausdorff issue with shuffle
`TestHausdorff.test_symmetry` and a number of other `directed_hausdorff` tests triggered the check for unknown objects we just put into `RandomState.shuffle` (see [here](https://github.com/scipy/scipy/pull/13463/checks?check_run_id=1799089671)):
```
mtrand.pyx:4478: in numpy.random.mtrand.RandomState.shuffle
    ???
E   UserWarning: `x` isn't a recognized object; `shuffle` is not guaranteed to behave correctly.
E.g., non-numpy array/tensor objects with view semantics may contain duplicates after shuffling.
```

### PR DESCRIPTION
**Pull request by rgommers**
Fixes CI failures and closes gh-13468.

### CODE REVIEW

**Review by Carreau — APPROVED** [review]

File: scipy/spatial/_hausdorff.pyx [hunk 1]
```diff
@@ -34,8 +36,12 @@ def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2, seed=0):
     rng = np.random.RandomState(seed)
     resort1 = np.arange(N1)
     resort2 = np.arange(N2)
-    rng.shuffle(resort1)
-    rng.shuffle(resort2)
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore",
+                                message="`x` isn't a recognized object")
+        rng.shuffle(resort1)
+        rng.shuffle(resort2)
```
Comment (tylerjereddy) [inline scipy/spatial/_hausdorff.pyx:43]: I'm just trying to understand--what's the danger of misbehavior here for data produced by `np.arange()`?

Or are you saying NumPy is likely going to change the behavior again (you mentioned special casing memoryviews) so this is a temporary work-around?

**Comment by rkern** [comment]
One solution here is just to use `cdef object resort1, resort2`. There's no real benefit to making `ndarray` objects with `arange()` only to convert them to `cdef long[:]`, then convert them back to `ndarray`s when you use them as fancy indices to other `ndarray`s.

**Comment by tylerjereddy** [comment]
Ok yeah, there may be no good reason for those specific typed memoryviews. I may have a tendency to type regardless to minimize "yellow lines" (alleged Python overhead) suggested by `cython -a` analysis.

That said, I'm not aware of any actual problem here, apart from the new NumPy suggesting that something might go wrong.

**Comment by rkern** [comment]
`cdef ndarray` is a good middle ground to make things `cython -a`-clean.

File: scipy/spatial/_hausdorff.pyx (hunk 1 above)
Comment (rgommers) [reply scipy/spatial/_hausdorff.pyx:43]: This is a 1-D array, so should be safe. But the code path it's taking is in general not safe for possibly non-contiguous array-likes. Hence it started to trigger a new warning in 1.20

**Comment by rgommers** [comment]
> `cdef ndarray` is a good middle ground to make things `cython -a`-clean.

Yes that's a better idea here than filtering, will update.

**Comment by rgommers** [comment]
Actually, change of plan - I want to merge this and change to `cdef ndarray` in a follow-up. It's hard to debug the multiple other CI failures with this one failing in a bunch of jobs.

**Comment by rgommers** [comment]
Follow-up in gh-13479.

**Pull request merged**